proposals/NNNN-objc-implementation.md

```diff
@@ -0,0 +1,323 @@
+* `@nonobjc` initializers must be convenience initializers, not designated or required initializers.
+
+> **Note**: `@objc @implementation` cannot support Swift-only designated and required initializers because subclasses with additional stored properties must be able to override designated and required initializers, but `@implementation` only supports overriding of `@objc` members. The Swift-only metadata that would be used for dynamic dispatch in an ordinary `@objc` class is not present in an `@implementation` class.
+
+### Member implementations
+
+Any non-`override` open, public, package, or internal `@objc` member of an `@objc @implementation extension` is a “member implementation”; that is, it implements some imported Objective-C member of the class it is extending. Member implementations are special because much of the compiler completely ignores them:
+
+* Access control denies access to member implementations in most contexts.
+* Module interfaces and generated interfaces do not include member implementations.
+* Objective-C generated headers do not include member implementations.
+
+This means that calls in expressions will *always* ignore the member implementation and use the imported Objective-C member instead. In other words, even other Swift code in the same module will behave as though the member is implemented in Objective-C.
+
+#### Rules
+
+A member implementation must:
+
+* Have the same Swift name as the member it implements.
+* Have the same selector as the member it implements.
+* Have the same foreign error convention and foreign async convention as the member it implements.
+* Not have other traits, like an overload signature, `@nonobjc`/`final` attribute, `class` modifier, or mutability, which is incompatible with the member it implements.
+* Not have `@_spi` attributes (they would be pointless since the visibility of the imported Objective-C attribute is what will make the member usable or not).
+
+Both the Swift name and the Objective-C selector of a member implementation must match the corresponding Objective-C declaration; Swift will diagnose an error if one matches but the other doesn't. This checking respects both the `@objc(custom:selector:)` in Swift implementations and the Swift name (`NS_SWIFT_NAME(custom(_:name:))`) attribute in Objective-C headers.
+
+Member implementations must have an overload signature that closely matches the Objective-C declaration’s. However, types that are non-optional in the Objective-C declaration may be implicitly unwrapped optionals in the member implementation if this is ABI-compatible; this is because Objective-C does not prevent clients from passing `nil` or implementations from returning `nil` when `nonnull` is used, and member implementations may need to implement backwards compatibility logic for this situation.
```

Comment:
Is there a different proposal which spells out the rules in more detail that we could link to? Or is this really the only difference of note?

Reply:
This is the only difference. Basically, the type match must be as close as a protocol witness would be, *except* that IUOs can be used in place of non-optional types if the two types would have exactly the same ABI.